lang/cs/Org.Apache.REEF.Common/Context/ContextConfiguration.cs

```diff
@@ -73,19 +74,37 @@ public sealed class ContextConfiguration : ConfigurationModuleBuilder
         [SuppressMessage("Microsoft.Security", "CA2104:Do not declare read only mutable reference types", Justification = "not applicable")]
         public static readonly OptionalImpl<IContextMessageHandler> OnMessage = new OptionalImpl<IContextMessageHandler>();
 
+        private static ConfigurationModule contextConfig;
+
+        private static readonly object ConfigLock = new object();
+
         public static ConfigurationModule ConfigurationModule
         {
             get
             {
-                return new ContextConfiguration()
-                    .BindNamedParameter(GenericType<ContextConfigurationOptions.ContextIdentifier>.Class, Identifier)
-                    .BindSetEntry(GenericType<ContextConfigurationOptions.StartHandlers>.Class, OnContextStart)
-                    .BindSetEntry(GenericType<ContextConfigurationOptions.StopHandlers>.Class, OnContextStop)
-                    .BindSetEntry(GenericType<ContextConfigurationOptions.ContextMessageSources>.Class, OnSendMessage)
-                    .BindSetEntry(GenericType<ContextConfigurationOptions.ContextMessageHandlers>.Class, OnMessage)
-                    .BindSetEntry(GenericType<TaskConfigurationOptions.StartHandlers>.Class, OnTaskStart)
-                    .BindSetEntry(GenericType<TaskConfigurationOptions.StopHandlers>.Class, OnTaskStop)
-                    .Build();
+                if (contextConfig == null)
+                {
+                    lock (ConfigLock)
+                    {
+                        if (contextConfig == null)
+                        {
+                            contextConfig = new ContextConfiguration()
```

Comment:
I am not sure `ComfigurationModules` are guaranteed to be immutable. If they are not, this code is dangerous. If they are immutable, this code, and the lock, can be replaced by a `private static readonly` field, and the `get()` here just returns a reference to it.

Reply:
I tried to create the configuration module and assign it to the static field, and only return the reference in get method. Interesting thing is it gives run-time error. 
With the current change, when multiple threads come to call the property together, with the lock, we will ensure it is created only once. And later, only the reference is returned. 
Can you illustrate why you think it might be dangerous?

Reply:
> I tried to create the configuration module and assign it to the static field, and only return the reference in get method. Interesting thing is it gives run-time error.

This is indeed interesting. From a cursory read of the classes `ConfigurationModule` and `ConfigurationModuleBuilder`, I don't see a reason for them to require creation outside of a `static` context. What was the error? I'm asking because a `static readonly` field is much more desirable than this construction with locks.

> Can you illustrate why you think it might be dangerous?

It would be dangerous if the returned object was mutable. However, `ConfigurationModule` is immutable and returns a deep copy of itself upon each `.Set()` call. Hence, I think the code as-is is safe. Still, just having a static field would be preferable.

Reply:
I tried it again to use statics but not able to repro. But I still clearly remember it gave me some error after I made the change. I will test more before changing it back ,

Reply:
oh I remember now, the error was because by the time I call ConfigurationModule property, the static variable initialization was not completed yet, causing crash. 
I believe the time that a static variable gets initialized varies, It fits to a simple variables. If it takes too long, it could be used before the initialization is done.

Reply:
>the error was because by the time I call ConfigurationModule property, the static variable initialization was not completed yet, causing crash.

According to [this description](https://blogs.msmvps.com/jonskeet/2010/01/26/type-initialization-changes-in-net-4-0/), it looks like we might have to define a `static` constructor in the `ConfigurationModuleBuilder` classes to get the behavior we want. However, that forces the initializer to always run, which might not be what we want.